android/jni/com/mapswithme/platform/SocketImpl.cpp

```diff
@@ -0,0 +1,75 @@
+#include "../core/jni_helper.hpp"
+#include "base/logging.hpp"
+#include "platform/socket.hpp"
+
+namespace platform
+{
+class SocketImpl : public Socket
+{
+public:
+  SocketImpl()
+  {
+    JNIEnv * env = jni::GetEnv();
+    g_socketClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/location/SocketWrapper");
+    static jmethodID const socketConstructor = jni::GetConstructorID(env, g_socketClass, "()V");
+    jni::TScopedLocalRef localSelf(env, env->NewObject(g_socketClass, socketConstructor));
+    g_self = env->NewGlobalRef(localSelf.get());
+    ASSERT(g_self, ());
+  }
+
+  ~SocketImpl()
+  {
+    Close();
+    JNIEnv * env = jni::GetEnv();
+    env->DeleteGlobalRef(g_self);
+    env->DeleteGlobalRef(g_socketClass);
+  }
+
+  bool Open(string const & host, uint16_t port)
+  {
+    JNIEnv * env = jni::GetEnv();
+    static jmethodID const openMethod =
+        jni::GetMethodID(env, g_self, "open", "(Ljava/lang/String;I)Z");
+    return env->CallBooleanMethod(g_self, openMethod, jni::ToJavaString(env, host),
+                                  static_cast<jint>(port));
+  }
+
+  void Close()
+  {
+    JNIEnv * env = jni::GetEnv();
+    static jmethodID const closeMethod = jni::GetMethodID(env, g_self, "close", "()V");
+    env->CallVoidMethod(g_self, closeMethod);
+  }
+
+  bool Read(uint8_t * data, uint32_t count)
+  {
+    JNIEnv * env = jni::GetEnv();
+    jbyteArray result = env->NewByteArray(count);
+    env->SetByteArrayRegion(result, 0, count, (const jbyte *)data);
```

Comment:
I guess you need to call read() on the Java side first, then call SetByteArrayRegion to copy from result to data.